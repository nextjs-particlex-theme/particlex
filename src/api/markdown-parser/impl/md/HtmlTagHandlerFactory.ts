import type HtmlTagHandler from '@/api/markdown-parser/impl/md/HtmlTagHandler'
import imageHandler from '@/api/markdown-parser/impl/md/handlers/ImageHandler'
import preHandler from '@/api/markdown-parser/impl/md/handlers/PreHandler'
import blockQuoteHandler from '@/api/markdown-parser/impl/md/handlers/BlockQuoteHandler'


const instanceHolder = new Map<string, HtmlTagHandler>()

/**
 * 注册一个 caster
 * @param caster caster
 */
const registerHandler = (caster: HtmlTagHandler) => {
  const tags = Array.isArray(caster.supportedTag) ? caster.supportedTag : [caster.supportedTag]
  for (const tag of tags) {
    const old = instanceHolder.get(tag)
    if (old) {
      throw new Error(`Duplicate MarkdownTagCaster, tag name: ${tag}`)
    }
    instanceHolder.set(tag, caster)
  }
}

registerHandler(imageHandler)
registerHandler(preHandler)
registerHandler(blockQuoteHandler)

/**
 * 包括一下，方便外面使用的时候好看.
 */
const wrapped = {
  getInstance: (tag: string): HtmlTagHandler | undefined => {
    return instanceHolder.get(tag)
  }
}

export default wrapped