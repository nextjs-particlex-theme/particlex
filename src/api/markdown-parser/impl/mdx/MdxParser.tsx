import type { MarkdownParser, ParsedMarkdown } from '@/api/markdown-parser/types'
import React from 'react'
import { compile, run } from '@mdx-js/mdx'
import * as runtime from 'react/jsx-runtime'
import os from 'node:os'
import MdxCodeBlock from '@/api/markdown-parser/impl/mdx/components/MdxCodeBlock'
import MdxImage from '@/api/markdown-parser/impl/mdx/components/MdxImage'
import MdxBlockQuote from '@/api/markdown-parser/impl/mdx/components/MdxBlockQuote'
import generateTocByMarkdown from '@/api/markdown-parser/common-toc-generator'
import createCommonHeadingWithId from '@/api/markdown-parser/impl/mdx/components/CommonHeadingWithId'

const components = {
  pre: MdxCodeBlock,
  img: MdxImage,
  blockquote: MdxBlockQuote,
  h1: createCommonHeadingWithId('h1'),
  h2: createCommonHeadingWithId('h2'),
  h3: createCommonHeadingWithId('h3'),
  h4: createCommonHeadingWithId('h4'),
  h5: createCommonHeadingWithId('h5')
}

async function parseMarkdownContent0(content: string): Promise<React.ReactNode> {
  const code = String(
    await compile(content, { outputFormat: 'function-body' })
  )
  // @ts-ignore
  const { default: MDXContent } = await run(code, {
    ...runtime,
    baseUrl: import.meta.url,
  })

  // Render the MDX content, supplying the ClientComponent as a component
  return <MDXContent components={components} />
}

type ParseError = Error & {
  line: number
  name: string
  column: number
  place: {
    line: number
    column: number
    offset: number
  }
}

function isParseAeeror(err: unknown): err is ParseError {
  if (!err || !(typeof err === 'object')) {
    return false
  }
  const errObj = err as Partial<ParseError>
  return !!errObj.column && !!errObj.line && !!errObj.name && !!errObj.place
}

function takeAboveLines(self: string[], lines: string[], startLine: number, lineCnt: number) {
  let cnt = Math.min(startLine, lineCnt - 1)

  for (let i = startLine - cnt; i <= startLine; i++) {
    self.push(` ${i.toString(10).padStart(4, ' ')} | ${lines[i]}`)
  }
}

/**
 * 处理 html 博客内容
 */
async function parseMarkdownContent(content: string): Promise<React.ReactNode> {
  try {
    return await parseMarkdownContent0(content)
  } catch (e) {
    if (!isParseAeeror(e)) {
      return Promise.reject(e)
    }
    // indicate where is wrong.
    const lines = content.split(os.EOL)
    const message = ['Failed to parse markdown content:']
    const programLine = e.line - 1

    message.push('')
    takeAboveLines(message, lines, programLine, 4)

    message.push('        ' + (new Array(e.column - 1).join(' ')) + '^')
    message.push('\t' + e.message)

    return Promise.reject(new Error(message.join(os.EOL), { cause: e }))
  }
}


const mdxParser: MarkdownParser = {
  async parse(markdown: string): Promise<ParsedMarkdown> {
    let node = await parseMarkdownContent(markdown)

    return {
      toc: generateTocByMarkdown(markdown),
      page: node
    }
  }
}

export default mdxParser
