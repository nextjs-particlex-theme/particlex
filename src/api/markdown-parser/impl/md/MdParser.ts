import type { MarkdownParser, ParsedMarkdown } from '@/api/markdown-parser/types'
import type React from 'react'
import reactParse, { Element } from 'html-react-parser'
import HtmlTagHandlerFactory from '@/api/markdown-parser/impl/md/HtmlTagHandlerFactory'
import { unified } from 'unified'
import remarkParse from 'remark-parse'
import remarkGfm from 'remark-gfm'
import adjustToc from '@/api/markdown-parser/common/toc-adjust-plugin'
import remarkRehype from 'remark-rehype'
import rehypeStringify from 'rehype-stringify'
import generateHeadingId from '@/api/markdown-parser/common/generate-heading-id-plugin'


const markdownToHtml = async (markdownContent: string): Promise<string> => {
  const parsed = await unified()
    .use(remarkParse)
    .use(remarkGfm)
    .use(adjustToc)
    .use(remarkRehype, { allowDangerousHtml: true })
    .use(generateHeadingId)
    .use(rehypeStringify, { allowDangerousHtml: true })
    .process(markdownContent)
  return String(parsed)
}

function processPostContent(html: string): React.ReactNode {
  return reactParse(html, {

    replace: (domNode) => {
      if (!(domNode instanceof Element)) {
        return
      }
      return HtmlTagHandlerFactory.getInstance(domNode.tagName)?.doCast(domNode)
    }
  })
}


const mdParser: MarkdownParser = {
  async parse(markdown: string): Promise<ParsedMarkdown> {
    const html = await markdownToHtml(markdown)
    const node = processPostContent(html)
    return Promise.resolve(node)
  }
} 

export default mdParser