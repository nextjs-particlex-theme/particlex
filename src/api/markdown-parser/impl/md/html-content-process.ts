import reactParse, { Element } from 'html-react-parser'
import type React from 'react'
import HtmlTagHandlerFactory from '@/api/markdown-parser/impl/md/HtmlTagHandlerFactory'


/**
 * 处理 html 博客内容
 */
export default function processPostContent(html: string): React.ReactNode {
  return reactParse(html, {
    replace: (domNode) => {
      if (!(domNode instanceof Element)) {
        return 
      }
      return HtmlTagHandlerFactory.getInstance(domNode.tagName)?.doCast(domNode)
    }
  })
}

