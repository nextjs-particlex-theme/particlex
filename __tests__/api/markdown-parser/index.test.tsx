import { render } from '@testing-library/react'
import parseMarkdown from '@/api/markdown-parser'
import md from './markdown.test.md'
import mdx from './markdown.test.mdx'
import '@testing-library/jest-dom'

test('Test md render', async () => {
  const parsed = await parseMarkdown(md, '.md')
  const rendered = render(parsed)
  expect(rendered.baseElement.innerHTML).toMatchSnapshot()
})

test('Test mdx render', async () => {
  const parsed = await parseMarkdown((mdx as unknown) as string, '.mdx')
  const rendered = render(parsed)
  expect(rendered.baseElement.innerHTML).toMatchSnapshot()
})