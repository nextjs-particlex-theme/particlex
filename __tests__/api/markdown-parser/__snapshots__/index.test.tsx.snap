// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test md render 1`] = `
"<div><p>Markdown是一种轻量级的「标记语言」。</p>
<h2 id="1-Markdown语法教程">1 Markdown语法教程</h2>
<h3 id="1.1-标题">1.1 标题</h3>
<p>不同数量的<code>#</code>可以完成不同的标题.。</p>
<h3 id="1.2-字体">1.2 字体</h3>
<p>粗体、斜体、粗体和斜体，删除线，需要在文字前后加不同的标记符号。如下：</p>
<p><strong>这个是粗体</strong></p>
<p><em>这个是斜体</em></p>
<p><em><strong>这个是粗体加斜体</strong></em></p>
<p><del>这里是删除线</del></p>
<p>注：如果想给字体换颜色、字体或者居中显示，需要使用内嵌HTML来实现。</p>
<h3 id="1.3-无序列表">1.3 无序列表</h3>
<p>无序列表的使用，在符号<code>-</code>后加空格使用。如下：</p>
<ul>
<li>无序列表 1</li>
<li>无序列表 2</li>
<li>无序列表 3</li>
</ul>
<p>如果要控制列表的层级，则需要在符号<code>-</code>前使用空格。如下：</p>
<ul>
<li>无序列表 1</li>
<li>无序列表 2
<ul>
<li>无序列表 2.1</li>
<li>无序列表 2.2</li>
</ul>
</li>
</ul>
<h3 id="1.4-有序列表">1.4 有序列表</h3>
<p>有序列表的使用，在数字及符号<code>.</code>后加空格后输入内容，如下：</p>
<ol>
<li>有序列表 1</li>
<li>有序列表 2</li>
<li>有序列表 3</li>
</ol>
<h3 id="1.5-引用">1.5 引用</h3>
<p>引用的格式是在符号<code>&gt;</code>后面书写文字。如下：</p>
<blockquote>
<p>读一本好书，就是在和高尚的人谈话。 ——歌德</p>
</blockquote>
<blockquote>
<p>雇用制度对工人不利，但工人根本无力摆脱这个制度。 ——阮一峰</p>
</blockquote>
<h3 id="1.6-链接">1.6 链接</h3>
<p><a href="https://github.com/nextjs-particlex-theme/particlex">particlex</a></p>
<h3 id="1.7-图片">1.7 图片</h3>
<p><img src="/favicon.ico" alt="favicon.ico" loading="lazy"></p>
<h3 id="1.8-分割线">1.8 分割线</h3>
<p>可以在一行中用三个以上的减号来建立一个分隔线，同时需要在分隔线的上面空一行。如下：</p>
<hr>
<h3 id="1.9-表格">1.9 表格</h3>
<p>可以使用冒号来定义表格的对齐方式，如下：</p>
<table><thead><tr><th align="left">姓名</th><th align="center">年龄</th><th align="right">工作</th></tr></thead><tbody><tr><td align="left">小可爱</td><td align="center">18</td><td align="right">吃可爱多</td></tr><tr><td align="left">小小勇敢</td><td align="center">20</td><td align="right">爬棵勇敢树</td></tr><tr><td align="left">小小小机智</td><td align="center">22</td><td align="right">看一本机智书</td></tr></tbody></table>
<h3 id="1.10-代码块">1.10 代码块</h3>
<p>如果在一个行内需要引用代码，只要用反引号引起来就好，如下：</p>
<p>Use the <code>printf()</code> function.</p>
<p>在需要高亮的代码块的前一行及后一行使用三个反引号，同时<strong>第一行反引号后面表示代码块所使用的语言</strong>，如下：</p>
<pre class="hljs"><div class="z-50 relative pt-12 pb-5"><div style="max-height: 50vh;"><span class="hljs-comment">// FileName: HelloWorld.java</span>
<span class="hljs-keyword">public</span> <span class="hljs-keyword">class</span> <span class="hljs-title class_">HelloWorld</span> {
  <span class="hljs-comment">// Java 入口程序，程序从此入口</span>
  <span class="hljs-keyword">public</span> <span class="hljs-keyword">static</span> <span class="hljs-keyword">void</span> <span class="hljs-title function_">main</span><span class="hljs-params">(String[] args)</span> {
    System.out.println(<span class="hljs-string">"Hello,World!"</span>); <span class="hljs-comment">// 向控制台打印一条语句</span>
  }
}
</div><div>java</div><div><span title="自动换行"><svg width="18" height="18"><use xlink:href="#wrap-line"></use></svg></span><span title="展开"><svg width="15" height="15"><use xlink:href="#fc-arrows-up-down"></use></svg></span><div title="复制"><svg width="15" height="15"><use xlink:href="#fc-copy"></use></svg></div></div></div></pre>
<pre class="hljs"><div class="z-50 relative pt-12 pb-5"><div style="max-height: 50vh;"><span class="hljs-tag">&lt;<span class="hljs-name">hello</span>&gt;</span>
	<span class="hljs-tag">&lt;<span class="hljs-name">world</span>&gt;</span><span class="hljs-tag">&lt;/<span class="hljs-name">world</span>&gt;</span>
<span class="hljs-tag">&lt;/<span class="hljs-name">hello</span>&gt;</span>
</div><div>xml</div><div><span title="自动换行"><svg width="18" height="18"><use xlink:href="#wrap-line"></use></svg></span><span title="展开"><svg width="15" height="15"><use xlink:href="#fc-arrows-up-down"></use></svg></span><div title="复制"><svg width="15" height="15"><use xlink:href="#fc-copy"></use></svg></div></div></div></pre>
<h2 id="2-其他语法">2 其他语法</h2>
<h3 id="2.1-HTML">2.1 HTML</h3>
<p>支持原生 HTML 语法，请写内联样式，如下：</p>
<p><span style="display: block; text-align: right; color: orangered;">橙色居右</span>
<span style="display: block; text-align: center; color: orangered;">橙色居中</span></p>
<h3 id="Github-警告">Github 警告</h3>
<blockquote><div class="flex items-center mt-4"><svg width="15" height="15"><use xlink:href="#icon-info"></use></svg><span class="ml-2">Note</span></div><p>Note 123</p>
</blockquote>
<blockquote><div class="flex items-center mt-4"><svg width="15" height="15"><use xlink:href="#fc-lightbulb"></use></svg><span class="ml-2">Tip</span></div><p>TIP 123</p>
</blockquote>
<blockquote><div class="flex items-center mt-4"><svg width="15" height="15"><use xlink:href="#fc-star"></use></svg><span class="ml-2">Important</span></div><p>IMPORTANT 123</p>
</blockquote>
<blockquote><div class="flex items-center mt-4"><svg width="17" height="17"><use xlink:href="#icon-warning"></use></svg><span class="ml-2">Warning</span></div><p>WARNING 123</p>
</blockquote>
<blockquote><div class="flex items-center mt-4"><svg width="15" height="15"><use xlink:href="#icon-info"></use></svg><span class="ml-2">Caution</span></div><p>CAUTION 123</p>
</blockquote></div>"
`;

exports[`Test mdx render 1`] = `
"<div><h2 id="Github-CodeBlock">Github CodeBlock</h2>
<h3 id="Examples">Examples</h3>
<blockquote><div class="flex items-center mt-4"><svg width="15" height="15"><use xlink:href="#icon-info"></use></svg><span class="ml-2">Note</span></div>
<p>NOTE</p>
</blockquote>
<blockquote><div class="flex items-center mt-4"><svg width="15" height="15"><use xlink:href="#fc-lightbulb"></use></svg><span class="ml-2">Tip</span></div>
<p>TIP</p>
</blockquote>
<blockquote><div class="flex items-center mt-4"><svg width="15" height="15"><use xlink:href="#fc-star"></use></svg><span class="ml-2">Important</span></div>
<p>IMPORTANT</p>
</blockquote>
<blockquote><div class="flex items-center mt-4"><svg width="17" height="17"><use xlink:href="#icon-warning"></use></svg><span class="ml-2">Warning</span></div>
<p>WARNING</p>
</blockquote>
<blockquote><div class="flex items-center mt-4"><svg width="15" height="15"><use xlink:href="#icon-info"></use></svg><span class="ml-2">Caution</span></div>
<p>CAUTION</p>
</blockquote>
<blockquote><div class="flex items-center mt-4"><svg width="15" height="15"><use xlink:href="#icon-info"></use></svg><span class="ml-2">Note</span></div>
<p> eee
hello world
[!NOTE]</p>
</blockquote>
<blockquote>
<p>hello [!INFO]
world</p>
</blockquote>
<blockquote><div class="flex items-center mt-4"><svg width="15" height="15"><use xlink:href="#icon-info"></use></svg><span class="ml-2">Note</span></div>
<p>NOTE</p>
</blockquote>
<h3 id="Notes">Notes</h3>
<blockquote>
<pre class="hljs"><div class="z-50 relative pt-12 pb-5"><div style="max-height: 50vh;">[!NOTE]
NOTE
</div><div>markdown</div><div><span title="自动换行"><svg width="18" height="18"><use xlink:href="#wrap-line"></use></svg></span><span title="展开"><svg width="15" height="15"><use xlink:href="#fc-arrows-up-down"></use></svg></span><div title="复制"><svg width="15" height="15"><use xlink:href="#fc-copy"></use></svg></div></div></div></pre>
</blockquote>
<blockquote><div class="flex items-center mt-4"><svg width="15" height="15"><use xlink:href="#icon-info"></use></svg><span class="ml-2">Caution</span></div>
<p>在 mdx 中必须使用 React 语法，例如 <code>class</code> 需要写成 <code>className</code>, <code>style</code>需要提供一个对象而不是直接的字符串。</p>
<p><strong>由于使用 SSR 渲染，所以无法绑定事件</strong>。</p>
</blockquote></div>"
`;
